src/main/scala/io/iohk/ethereum/consensus/Consensus.scala

```diff
@@ -1,41 +1,187 @@
+  ): Either[ValidationBeforeExecError, BlockExecutionSuccess]
+
   /**
-   * Provides the [[io.iohk.ethereum.validators.BlockHeaderValidator BlockHeaderValidator]] that is specific
-   * to this consensus protocol.
+   * This function validates that the various results from execution are consistent with the block. This includes:
+   *   - Validating the resulting stateRootHash
+   *   - Doing BlockValidator.validateBlockReceipts validations involving the receipts
+   *   - Validating the resulting gas used
+   *
+   * @note This method was originally provided by the [[io.iohk.ethereum.ledger.Ledger Ledger]].
+   *
+   * @param block to validate
+   * @param stateRootHash from the resulting state trie after executing the txs from the block
+   * @param receipts associated with the execution of each of the tx from the block
+   * @param gasUsed, accumulated gas used for the execution of the txs from the block
+   * @return None if valid else a message with what went wrong
    */
-  // FIXME Probably include the whole of [[io.iohk.ethereum.validators.Validators]].
-  def blockHeaderValidator: BlockHeaderValidator
+  def validateBlockAfterExecution(
+    block: Block,
+    stateRootHash: ByteString,
+    receipts: Seq[Receipt],
+    gasUsed: BigInt
+  ): Either[BlockExecutionError, BlockExecutionSuccess]
+}
+
+/**
+ * Internal API, used for testing.
+ *
+ * This is a [[Consensus]] API for the needs of the test suites.
+ * It gives a lot of flexibility overriding parts of a consensus' behavior
+ * but it is the developer's responsibility to maintain consistency (though the
+ * particular consensus protocols we implement so far do their best
+ * in that direction).
+ */
+trait TestConsensus extends Consensus {
+  def blockGenerator: TestBlockGenerator
+
+  /** Internal API, used for testing */
+  protected def newBlockGenerator(validators: Validators): TestBlockGenerator
 
-  // Ledger uses this in importBlock
-  def validateBlockBeforeExecution(block: Block): Either[ValidationBeforeExecError, BlockExecutionSuccess] = ???
+  /** Internal API, used for testing */
+  def withValidators(validators: Validators): TestConsensus
+
+  /** Internal API, used for testing */
+  def withVM(vm: VMImpl): TestConsensus
+
+  /** Internal API, used for testing */
+  def withBlockGenerator(blockGenerator: TestBlockGenerator): TestConsensus
 }
+
+abstract class ConsensusImpl[C <: AnyRef](
+  theVm: VMImpl,
+  blockchain: BlockchainImpl,
+  blockchainConfig: BlockchainConfig,
+  fullConsensusConfig: FullConsensusConfig[C]
+) extends TestConsensus with Logger {
+
+  final type Config = C
+
+  protected val _blockPreparator = new BlockPreparator(
+    vm = vm,
+    signedTxValidator = validators.signedTransactionValidator,
+    blockchain = blockchain,
+    blockchainConfig = blockchainConfig
+  )
+
+  /**
+   * This is used by the [[io.iohk.ethereum.consensus.Consensus#blockGenerator blockGenerator]].
+   */
+  def blockPreparator: BlockPreparator = this._blockPreparator
```

Comment:
why the extra val? can't it just be `override val blockPreparator = new BlockPreparetor(..)`?

Reply:
There is no extra `val`, just one: `_blockPreparator`.

Reply:
And that's the one I'm referring to :)
Why not:
```
override val blockPreparator: BlockPreparator = new BlockPreparator(
    vm = vm,
    signedTxValidator = validators.signedTransactionValidator,
    blockchain = blockchain,
    blockchainConfig = blockchainConfig
  )
```
instead of
```
  protected val _blockPreparator = new BlockPreparator(
    vm = vm,
    signedTxValidator = validators.signedTransactionValidator,
    blockchain = blockchain,
    blockchainConfig = blockchainConfig
  )
  def blockPreparator: BlockPreparator = this._blockPreparator
```

Reply:
Ok, nvm, I see why :) This class is meant to be inherited, we want to keep it a def on this level.

Reply:
Well, I don't get it. `ConsensusImpl` already inherits those defs from `Consensus`, what's the point of implementing them here rather than:

```
class AtomixRaftConsensus private(
  val vm: VMImpl,
  val blockchain: BlockchainImpl,
  val blockchainConfig: BlockchainConfig,
  val fullConsensusConfig: FullConsensusConfig[AtomixRaftConfig],
  val validators: Validators,
  val blockGenerator: AtomixRaftBlockGenerator
) extends ConsensusImpl[AtomixRaftConfig] ...
```

Also, as Lukasz noted above, why the pair:

```
protected val _blockPreparator = ...
def blockPreparator: BlockPreparator = ...
```

instead of one:

```
val blockPreparator: BlockPreparator = ...
```

These are not big issues, but it seems they unnecessarily increase the volume of code. Is there a purpose for these constructions that I'm missing?